### PR TITLE
feat(ui-concerto) : relationship provider

### LINKED ISSUE #201: Ability to use drop-downs to specify relationships in form
# Feature Request 🛍️
Concerto Form currently supports editing relationships, but they are edited as free text. It would be a much better editing experience to be able to select the target of a relationship using a drop-down or search component.

## Use Case
Editing relationships as free text input fields is error prone and confusing for end users.

## Possible Solution
Allow the application developer to pass a `RelationshipProvider` option to `ConcertoForm`. The form calls a `getOptions(field)` method to resolve the relationship targets for a given field.

## Context
Useful to support signature blocks.

## Detailed Description
A new option for Concerto Form: `relationshipProvider` which is an object with a `getOptions` method, which takes a Concerto `Field` as its sole arguments, and returns an array of objects with `key, value, text` properties.

### PR DESCRIPTION
**Pull request by dselman**
Signed-off-by: Dan Selman <danscode@selman.org>

# Closes #201 
Adds a relationship provider to Concerto Form.

### Changes
- Add an new optional parameter to allow relationships to populate a drop-down.

### Screenshots or Video

image

### CODE REVIEW

**Review by DianaLease — COMMENTED** [review]
This is awesome! The only thing I'm confused about is how the dropdown is initialized. When I first load it in the storybook before selecting a value, I see `resource:test.Person#0660` as the value of owner, which I don't really understand.

See below the value of `owner` before selection (`resource:test.Person#0660`) and after selection (`test.Person#Marissa`).
Screen Shot 2020-09-28 at 3 25 58 PM

File: packages/storybook/src/stories/3-ConcertoForm.stories.js [hunk 1]
```diff
@@ -55,6 +60,42 @@ export const Demo = () => {
     }
   };
 
+  options.relationshipProvider = {
+    getOptions : (field) => {
+      if(field.getFullyQualifiedTypeName() === 'test.Person') {
```
Comment (mttrbrts) [inline packages/storybook/src/stories/3-ConcertoForm.stories.js:65]: Why is this necessary? Is field `null` on an early render, perhaps?

**Review by mttrbrts — APPROVED** [review]

File: packages/storybook/src/stories/3-ConcertoForm.stories.js (hunk 1 above)
Comment (dselman) [reply packages/storybook/src/stories/3-ConcertoForm.stories.js:65]: No, I just wanted to illustrate how to define a relationship provider that is sensitive to the type of the relationship.

**Comment by dselman** [comment]
> This is awesome! The only thing I'm confused about is how the dropdown is initialized. When I first load it in the storybook before selecting a value, I see `resource:test.Person#0660` as the value of owner, which I don't really understand.
> 
> See below the value of `owner` before selection (`resource:test.Person#0660`) and after selection (`test.Person#Marissa`).
> Screen Shot 2020-09-28 at 3 25 58 PM

Good point. I'm not sure how we should handle initialisation. Theoretically we could delegate to the relationship providers to provide a default value - but that feels a bit heavy. I suggest we leave it as-is for now and revisit if/when more requirements emerge (e.g. supporting a query interface).
